ocaml/typing/typecore.ml

```diff
@@ -285,48 +285,40 @@ let case lhs rhs =
 
 type position_in_function = FTail | FNontail
 
-
 type position_in_region =
-  (* not the tail of a region*)
   | RNontail
-  (* tail of a region,
-     together with the mode of that region,
-     and whether it is also the tail of a function
-     (for tail call escape detection) *)
+  (** not the tail of a region*)
+
   | RTail of Value_mode.t * position_in_function
+  (** tail of a region, together with the mode of that region, and whether it is
+     also the tail of a function (for tail call escape detection) *)
 
 type expected_mode =
   { position : position_in_region;
-    escaping_context : Env.escaping_context option;
-    (* the upper bound of mode*)
-    mode : Value_mode.t;
-    (* in some scnearios, the above `mode` will be the exact mode of the
-        expression to be typed, indicated by the `exact` field.
-
-    - In any case, there is no risk of miscompilation in taking an upper bound
-    as exact. We might lose some range and trigger some false mode errors.
+    (** position in the current region  *)
 
-    - Taking an exact as upper bound could cause issues. In particular
-    for the inner function of an uncurried function.
+    escaping_context : Env.escaping_context option;
+    (** explains why [mode] is low. Better be empty if [mode] is max *)
 
+    mode : Value_mode.t;
+    (** the upper bound *)
 
-    Therefore, if we just take it as exact regardless of the `exact`
-    field, we should be safe. Moreover, note that for most allocations, they
-    want to use expected_mode.mode as exact anyway, because that would be the
-    only constraint and they want to be as local as possible. The only exception
-    is uncurried functions where the mode constraints are tricky.
-    *)
     exact : bool;
+    (** if [true], means [mode] is exact (not just upper bound) *)
+
     tuple_modes : Value_mode.t list;
-    (* for t in tuple_modes, t <= regional_to_global mode *)
+    (** User experience improvement. For t in tuple_modes, t <=
+    regional_to_global mode. Always safe to set it to empty (and thus
+    disabling this improvement) *)
```

Comment:
Thanks - that does sound better. Speaking of length mismatch, that sounds like a compiler BUG if it happens. Should we `assert false` in that case?

Also maybe `tuple_modes : Value_mode.t list option` is a better type?

Reply:
I'd rather just ignore it and not have to worry about whether the arity has already been type-checked or not. It would indeed also push us to use `list option` and I'd rather not bother.

Reply:
I think it's not much work - can I spend 10 mins to fix it to feel safer?